### ISSUE TITLE
v5 new components: BottomBar

### ISSUE THREAD

**Issue by ss-dimasm** [reapit/elements#224]
**Abstract**

While the smaller part of the `BottomBar` is live in main (#192), we could start the next step by scaffolding the BottomBar based on the design spec. Further technical discussion is required before we proceed with crafting the production-grade implementation

**Specification**
- Is new component or existing? new
- Is the component being re-named? no
- Is the component design signed off by product and design? yes
- [Link to DS in Figma](https://www.figma.com/design/XJ6qcAV8gHscsUodqJMNEF/Reapit-Elements-production-ready-components?node-id=34-7016&node-type=instance&m=dev)
- [Link to Design Guidelines](https://www.figma.com/design/pJ96bxoEFStjjIbAPy0Jaj/Design-Guidelines?node-id=945-25700&node-type=symbol&m=dev)

**Developer Checklist**

- [ ] Styles alignment between Design System and Elements
- [ ] Check design tokens in Figma and implement CSS variable tokens if available for relevant component
- [ ] Align with accessibility standards / spec as per above
- [ ] If relevant, break down component into `Styles Only` and `React` component structures
- [ ] Ensure all variants of components are documented as appropriate 
- [ ] Ensure unit test coverage is adequate for component
- [ ] Update documentation in MDX file as per guidelines
- [ ] Changelog updated to reflect a single beta version per component ideally

**Release Checklist**

- [ ] Approved PR merged to main
- [ ] Design & product review and feedback addressed by developer
- [ ] Beta release by product / engineering lead to next beta version

**Additional Context or Information**
- The `BottomBar` should be hidden when users scroll down and should appear when users scroll up [Detail Link](https://www.figma.com/design/XJ6qcAV8gHscsUodqJMNEF/Reapit-Elements-production-ready-components?node-id=34-7016&node-type=instance&m=dev)
- When the expanded menu is visible and the user scrolls down, the panel should close, and the bar should be hidden [Question Link](https://www.figma.com/design/XJ6qcAV8gHscsUodqJMNEF?node-id=34-7014&m=dev#1040987505)

**Comment by ss-dimasm** [comment]
Hi @kurtdoherty and @adrian-reapit , FYI I've raised this ticket to initiate the investigation of BottomBar implementation. I will provide updates in this ticket and create a new bare minimum proof-of-concept (POC) first. Then, I'll share the findings with you, including considerations for the next steps thanks 👍

**Comment by ss-dimasm** [comment]
Hello @kurtdoherty and @adrian-reapit , the BottomBar POC looks okay for now

https://github.com/user-attachments/assets/db3f7ef9-98b3-4729-b3e8-5996f88e9b28
I think we could 

In terms of production grade API, I think we could achieve this behavior 
```tsx
<BottomBar scrollingState={scrollingState}>
  <BottomBar.Item icon={<Icon icon="star" />} href="#" isActive>A</BottomBar.Item>
  <BottomBar.Item icon={<Icon icon="star" />} href="#" hasBadge>B</BottomBar.Item>
  <BottomBar.Item icon={<Icon icon="star" />} href="#">C</BottomBar.Item>
  <BottomBar.Item icon={<Icon icon="star" />} href="#">D</BottomBar.Item>
  <BottomBar.Item icon={<Icon icon="star" />} href="">E</BottomBar.Item>
  <BottomBar.Menu>
    {({ open }) => (
      {open ? (
        <BottomBar.Menu.Item icon={<Icon icon="star" />} href="#">F</BottomBar.Menu.Item>
        <BottomBar.Menu.Item icon={<Icon icon="star" />} href="#">G</BottomBar.Menu.Item>
       : null
    )}
  </BottomBar.Menu>
</BottomBar>
```

**Comment by kurtdoherty** [comment]
**suggestion:** It's not clear what `scrollingState` is intended for, why it needs to be passed in by the consumer and what values the prop accepts. Can you please explain that some more?

**question:** What is the need for `BottomBar.Menu` and why can't we simply use the existing `Menu` component with a `BottomBar.Item` as the trigger?

**Comment by ss-dimasm** [comment]
Hey @kurtdoherty , sorry my last comment is accidentally pressed, I've update that would you mind to recheck again?

**Comment by kurtdoherty** [comment]
> having a scrollingState prop makes the component more widely usable.

**question:** how does this make it "more widely usable"? This is a very general claim and you haven't mentioned any specific use-cases that your thinking of where this prop would be useful. Would also be helpful if you explain the different kinds of values this prop could be given. Is it a boolean? An enum? Something else?

> an upper layer of Menu that would make easier integration with scrollingState from users perspective

I don't know what this means... I also have the following questions re this `BottomBar.Menu` component:

**question:** What is the purpose of the render-prop child for `BottomBar.Menu` and its `open` param? That's not how our current `Menu` component works, so I'm not sure why it's needed here.

**question:** I'm not sure I understand why the BottomBar.Menu.Item accepts an icon? I don't see any icons in use for the menu items in the designs...

**note:** I've been having some conversation with Andrei in the Figma discussion thread where you discussed the need to close the menu. Would be worth reading that as I think there's been some valuable clarifications come out of the extra comments.

> BottomBar.Menu.Item

**suggestion:** I'd recommend avoiding three levels of nesting. Seems like there's no reason consumers can just use `Menu.Item` directly, right?

**Comment by ss-dimasm** [comment]
> how does this make it "more widely usable"? This is a very general claim and you haven't mentioned any specific use-cases that your thinking of where this prop would be useful

Users could decide whether they would use the slides in-out behavior; so it would stick at the bottom if they don't want to have that. The `scrollingState` could be marked as the container scroll position (Y) and also an undefined. I am planning to generate the `scrollingState` props from the container's scroll listener hook that could indicate is user scrolling up or down

> Would also be helpful if you explain the different kinds of values this prop could be given. Is it a boolean? An enum? Something else?

Apologies my previous comment is not clear enough, the `scrollingState` would access an object `{isScrollingUp: boolean, isScrollingDown: boolean}` or an undefined, if it's an object then the slides in-out behavior would applied otherwise it would make the bar hanging at the bottom

> What is the purpose of the render-prop child for BottomBar.Menu and its open param? That's not how our current Menu component works, so I'm not sure why it's needed here

Yes, so the `BottomBar` is actually have a Menu when it's has an expandable state, the fifth icon is the entry point for that, and to make it more easier from user perspective, I am planning on make a wrapper that wrap the `Menu`. This would make more easier to integrate with the slides in-out behavior that actually comes from `BottomBar`, we could ignore the `open` param for now but it's designed to simplified the usage at the end

I am thinking on we could achieve this one pattern from users perspective
```tsx
<BottomBar>
  ... rest
  <BottomBar.Menu icon={<Icon icon='more' />}>
  {({getTriggerProps}) => (
   <BottomBarItem icon={<Icon icon='foo' />}>Menu 6</BottomBarItem>
  )}
  </BottomBar.Menu />
</BottomBar>
```
Instead of...

```tsx
<BottomBar>
  <Menu data-alignment="right">
       <Menu.Trigger>
            {({ getTriggerProps }) => (
              <BottomBarItem
                onClick={() => "test"}
                {...getTriggerProps()}
                icon={<AiOutlineEllipsis size={18} />}
                hasBadge
              >
                More
              </BottomBarItem>
            )}
          </Menu.Trigger>
          <Menu.Popover yOffset={-15}>
            <Menu.List>
              <Menu.Item onClick={console.log}>Button 5</Menu.Item>
              <Menu.Item onClick={console.log}>Button 6</Menu.Item>
              <Menu.Item closeMenu={false}>Button 7</Menu.Item>
            </Menu.List>
          </Menu.Popover>
        </Menu>
</BottomBar>
```

> I'm not sure I understand why the BottomBar.Menu.Item accepts an icon? I don't see any icons in use for the menu items in the designs... I assume this is just a copy-paste error in the code snippet

It's a typo once again apologies, it should be `BottomBarItem` component or `BottomBar.Item` (just an alias from `BottomBarItem`; This component already live in the `main` branch

> I'd recommend avoiding three levels of nesting. Seems like there's no reason consumers can just use Menu.Item directly, right?

Same as above

>  I've been having some conversation with Andrei in the Figma discussion thread where you discussed the need to close the menu. Would be worth reading that as I think there's been some valuable clarifications come out of the extra comments.

Thanks for that, I'll take a look

**Comment by ss-dimasm** [comment]
> I've been having some conversation with Andrei in the Figma discussion thread where you discussed the need to close the menu. Would be worth reading that as I think there's been some valuable clarifications come out of the extra comments

This is what I've been considered, current `Menu` is allowing us to scroll down once it's opened; but with the latest conversation (lock scrolling ... while it's open), we might need a new layer in top of the `Menu` and above the `BottomBar` that act to prevent to scrolling 🤔

**Comment by kurtdoherty** [comment]
Thanks for the responses 🙏 

### Re: scrollingState

**suggestion:** This feels like a futurism, and not the kind of futurism that we should pursue right now. There's two kinds of flexibility we will often be faced with in our work: (1) flexibility for different atoms to be composed together to obtain a different UI, and (2) flexibility for a specific atom's behaviour to be adjusted. What you're suggesting with `scrollingState` feels like (2). When I talk about decoupling atoms in issue and PR feedback, I'm (almost always) advocating for (1).

Unless Design have expressed a need for us to disable the show/hide behaviour of the BottomBar atom, that's not something we should design for in our interface. It's sufficient for us to know that it's a behaviour of the BottomBar atom, so if we need to be able to disabled or change that behaviour in future, it will be BottomBar that will be modified.

### Re: BottomBar and the "More" menu item

**issue:** I agree that there's an opportunity here to encapsulate a good chunk of the composition required for a menu and it's items, but I don't understand either approach you've now suggested.

In the first instance, your proposal used a render-prop like `({ open }) => ...` to conditionally render the menu _items_. In the second instance, you're now using the Menu.Trigger's render-prop (which provides `getTriggerProps`), but there's no ability to specify the menu _items_. In both cases, it feels like you're getting confused about what can be encapsulated here.

**suggestion:** I think what you might be looking for is this...

```jsx
<BottomBar>
  ...
  <BottomBar.MoreMenu>
    <Menu.Item>Item 1</Menu.Item>
    <Menu.Item>Item 2</Menu.Item>
    <Menu.Item>Item 3</Menu.Item>
  </BottomBar.MoreMenu>
</BottomBar>
```

Here, the `BottomBar.MoreMenu` compound component _encapsulates_ the composition of `Menu`, `Menu.Trigger`, `Menu.Popover` and `Menu.List`, in addition to the specific icon and label text used for the Menu's trigger (which would just be another `BottomBar.Item`. That is, `BottomBar.MoreMenu` would effectively be:

```javascript
function BottomBarMoreMenu({ children }) {
  return (
    <Menu>
      <Menu.Trigger>
        {({ getTriggerProps }) => <BottomBarMoreItem {...getTriggerProps()} />}
      </Menu.Trigger>
      <Menu.Popover yOffset={-15}>
        <Menu.List>
          {children}
        </Menu.List>
      </Menu.Popover>
    </Menu>
  )
}
```

**note:** This kind of encapsulation is only possible if Design confirm that the "More" item in a BottomBar will always be exactly as they've designed it. If the "More" item in different products is likely to use a different icon or label text, there's not much point doing this encapsulation in Elements.

### Re: Scroll blocking for open Menus

**note:** Now that Andrei has expressed this requirement for Menu's, it is the `Menu` atom that will need to implement this behaviour, not some wrapper for Menu in `BottomBar`. This is good, because it means BottomBar should not need to care whether the "More" menu is open or not; it can just respond to scrolling as it needs and the Menu component will take care of ensuring scrolling is blocked no matter where it's used in the product.

Until this change is implemented in Menu, `BottomBar` will hide itself whether the "More" menu is open or not.

**Comment by ss-dimasm** [comment]
> here's two kinds of flexibility we will often be faced with in our work: (1) flexibility for different atoms to be composed together to obtain a different UI, and (2) flexibility for a specific atom's behaviour to be adjusted

Can we also make an atom that actually present the same UI output? the first atom could always have the slides in-out behavior and the second it would stick hanging in the bottom of the view port. Both could be achieved once we compose it

```tsx
// default with the animation
<ElBottomBar> ... </ElBottomBar>
// sticky variant
<ElBottomBarSticky> ... </ElBottomBarSticky>
```

> This kind of encapsulation is only possible if Design confirm that the "More" item in a BottomBar will always be exactly as they've designed it

At first, thanks for the suggested interfaces really appreciate it and I completely agree that's what I want to achieve for production-grade of the `BottomBar`. I think we could extend it more by accepting the some props in `BottomBarMoreMenu` as a setting for the `BottomBarItem`, since based on the spec the entry point of the `Menu` is always formed with the `BottomBarItem`

```tsx
// settings property is the same as `BottomBarItemProps` but it's deliberately passed to the entry point that trigger the `Menu`
function BottomBarMenu({ children, settings}) {
  return (
    <Menu>
      <Menu.Trigger>
        {({ getTriggerProps }) => <BottomBarItem  {...settings} {...getTriggerProps()} />} // proposed
      </Menu.Trigger>
      <Menu.Popover yOffset={-15}>
        <Menu.List>
          {children}
        </Menu.List>
      </Menu.Popover>
    </Menu>
  )
}
```
By doing this we could let the consumer decide the entry point settings such like which icon and label that they should have, what do you think? but at the end we still need to confirm it with the Design Team

> Until this change is implemented in Menu, BottomBar will hide itself whether the "More" menu is open or not

This will simplified the focus, thanks for this

**Comment by kurtdoherty** [comment]
> Can we also make an atom that actually present the same UI output?

Yes, we could take that approach, but we don't need to right now.

> I think we could extend it more by accepting the some props in BottomBarMoreMenu

Yes, we can expand the interface in a number of ways if we need to provide some targeted customisation. I don't believe we'll need to do this though.

**note:** Just in case you didn't notice it, my code snippet involves two new compound components: `BottomBarMoreItem` and `BottomBarMoreMenu`. The first one encapsulates the specific label and icon expected of the "More" item, while the latter encapsulates the composition of the menu components, including the `BottomBarMoreItem`.

**Comment by ss-dimasm** [comment]
>  Just in case you didn't notice it, my code snippet involves two new compound components: BottomBarMoreItem and BottomBarMoreMenu

yes I notice the new compound components, so I guess we could scaffold the `BottomBar` with this interface

```tsx
<BottomBar>
 <BottomBarItem icon={<Icon icon="star" />} isActive>
   Menu 1
 </BottomBarItem>
 <BottomBarItem icon={<Icon icon="star" />}>
   Menu 2
 </BottomBarItem>
 <BottomBarItem icon={<Icon icon="star" />}>
   Menu 3
 </BottomBarItem>
 <BottomBarItem icon={<Icon icon="star" />} hasBadge>
   Menu 4
 </BottomBarItem>
 <BottomBar.MoreMenu>
   <Menu.Item>Menu 6</Menu.Item>
   <Menu.Item>Menu 7</Menu.Item>
 </BottomBar.MoreMenu>
</BottomBar>
```

the `BottomBar` would act as the wrapper around the regular `BottomBarItem` and it would provide the new API (`MoreMenu`) which is the output of the encapsulation around `Menu` component

> my code snippet involves two new compound components: `BottomBarMoreItem` and `BottomBarMoreMenu`. The first one encapsulates the specific label and icon expected of the "More" item
 
the encapsulation is good here and I agree for the `BottomBarMoreMenu`. However, for the `BottomBarMoreItem`, we might need to raise a new question to the Design Team; can the "more menu" have a badge? And should that component be able to receive the `isActive` state?

assuming we're on page 7 and the it's registered as part of the `BottomBarMoreMenu`, should we change the focus or "active" state to the `BottomBarMoreMenu` or will it lose focus once we're on the seventh page?

**Comment by ss-dimasm** [comment]
> can the "more menu" have a badge? And should that component be able to receive the isActive state

**note:** initial query raised to design team https://www.figma.com/design/XJ6qcAV8gHscsUodqJMNEF/Reapit-Elements-production-ready-components?node-id=34-6905&node-type=instance&m=dev

**Comment by kurtdoherty** [comment]
> can the "more menu" have a badge?
> assuming we're on page 7 and the it's registered as part of the BottomBarMoreMenu, should we change the focus or "active" state

Both of these are great questions! 👏 

For the second one, I believe Andrei and I resolved that [here](https://www.figma.com/design/XJ6qcAV8gHscsUodqJMNEF?node-id=15-3629&m=dev#1023243072). Technically, that was for a main nav item with a menu, but I imagine the answer would be the same for the mobile nav "more" item.

**Comment by ss-dimasm** [comment]
Thanks for the details @kurtdoherty , I think I could start shipping the component tomorrow based on our conversation above, or we might need to wait until the dependencies are resolved first? (the queries that I've asked)

**Comment by kurtdoherty** [comment]
Nah, no need to wait. I don't think what we've discussed will need to change drastically.

**Comment by andrei-dragut** [comment]
@ss-dimasm Looks good. Only one comment:
- the overflow menu item (More) needs to have the same width as the other items. All items must share the bar width equally.

**Comment by ss-dimasm** [comment]
thanks @andrei-dragut for spot this one, actually we've our conversation last year about this one https://github.com/reapit/elements/pull/233#discussion_r1870390469, and we're agreed to not override another instance component style (Menu in this case) to match with the spec

but given with current situation, seems it's worthy to amend the existing bottom bar behavior itself without touching another scope; FYI @kurtdoherty I've raise new PR for this one it's on https://github.com/reapit/elements/pull/296 thanks

**Comment by kurtdoherty** [comment]
@ss-dimasm I don't think the [conversation in #233](https://github.com/reapit/elements/pull/233#discussion_r1870390469) is relevant to Andrei's comment; our `BottomBar.MoreMenu` component is simply missing some important styles to ensure the div rendered by the Menu component used internally grows in the flex container like the other `BottomBar.Item` components do 👍
